staging/src/k8s.io/apiserver/pkg/endpoints/filters/request_deadline_test.go

```diff
@@ -438,6 +443,961 @@ func TestWithFailedRequestAudit(t *testing.T) {
+		if !ok {
+			t.Errorf("expected ResponseWriter object to implement FlushError")
+			return
+		}
+
+		ctrl := http.NewResponseController(w)
+		if err := ctrl.SetWriteDeadline(time.Now().Add(100 * time.Millisecond)); err != nil {
+			t.Errorf("expected no error from SetReadDeadline, but got: %v", err)
+			return
+		}
+
+		// a) the handler blocks here indefinitely
+		// b) this forces the handler write timeout
+		// to occur on the server side
+		// c) the http2 client should receive a stream error
+		// immediately after the write timeout occurs.
+		<-clientDoneCh
+
+		// Write to the ResponseWriter object may succeed immediately
+		// after the timeout occurs if the underlying buffer has room.
+		if _, err := w.Write([]byte("hello")); err != nil {
+			t.Logf("ResponseWriter.Write returned an error after timeout: %v", err)
+		}
+		// FlushError, on the other hand, is expected to
+		// return a timeout error immediately.
+		if err := flusher.FlushError(); !errors.Is(err, os.ErrDeadlineExceeded) {
+			t.Errorf("expected an os.ErrDeadlineExceeded, but got: %v", err)
+		}
+
+		// NOTE: the handler goroutine is expected to terminate as soon
+		// as it receives an error from either Write or FlushError. This
+		// is how we prevent a request handler from running indefinitely.
+	}))
+
+	defer server.Close()
+	server.EnableHTTP2 = true
+	server.StartTLS()
+
+	client := server.Client()
+	func() {
+		defer close(clientDoneCh)
+		_, err := client.Get(server.URL + "/foo")
+		if !responsewritertesting.IsStreamReadOrWriteTimeout(err) {
+			t.Errorf("expected a stream reset error, but got: %v", err)
+		}
+	}()
+
+	select {
+	case <-handlerDoneCh:
+	case <-time.After(wait.ForeverTestTimeout):
+		t.Errorf("expected the request handler to have terminated")
+	}
+}
+
+func TestPerHandlerWriteDeadlineHTTP1WithTimeoutBeforeHandlerWrites(t *testing.T) {
+	// This test documents the behavior of the per handler write
+	// deadline with a standard net http/1x server.
+	//
+	// scenario: http/1x, write deadline is set, timeout occurs before
+	// the handler writes to the ResponseWriter object.
+	t.Parallel()
+	handlerDoneCh := make(chan struct{})
+	server := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		defer close(handlerDoneCh)
+		if req.ProtoMajor != 1 {
+			t.Errorf("expected an HTTP/1x request, but got: %s", req.Proto)
+			return
+		}
+		flusher, ok := w.(interface{ FlushError() error })
+		if !ok {
+			t.Errorf("expected ResponseWriter object to implement FlushError")
+			return
+		}
+
+		ctrl := http.NewResponseController(w)
+		if err := ctrl.SetWriteDeadline(time.Now().Add(1 * time.Millisecond)); err != nil {
+			t.Errorf("expected no error from SetReadDeadline, but got: %v", err)
+			return
+		}
+
+		// NOTE: if we block the handler here, the client (not setting a
+		// transport or client timeout) will hang indefinitely, even
+		// though we have set the write deadline on the server side.
```

Comment:
self: open an issue with the golang team

Reply:
The description at https://pkg.go.dev/net/http@go1.21.6#ResponseController.SetWriteDeadline is consistent with the behavior you report: the deadline is not a deadline for some (something, anything), it is _only_ a tripwire for write operations that happen after the deadline.